stripe/src/main/java/com/stripe/android/view/CardNumberEditText.java

```diff
@@ -0,0 +1,175 @@
+package com.stripe.android.view;
+
+import android.content.Context;
+import android.support.annotation.VisibleForTesting;
+import android.text.Editable;
+import android.text.InputFilter;
+import android.text.TextWatcher;
+import android.util.AttributeSet;
+import android.widget.EditText;
+
+import com.stripe.android.model.Card;
+import com.stripe.android.util.StripeTextUtils;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * An {@link EditText} that handles spacing out the digits of a credit card.
+ */
+public class CardNumberEditText extends EditText {
+
+    private static final int MAX_LENGTH_COMMON = 19;
+    // Note that AmEx and Diners Club have the same length
+    // because Diners Club has one more space, but one less digit.
+    private static final int MAX_LENGTH_AMEX_DINERS = 17;
+
+    private static final Integer[] SPACES_ARRAY_COMMON = {4, 9, 14};
+    private static final Set<Integer> SPACE_SET_COMMON =
+            new HashSet<>(Arrays.asList(SPACES_ARRAY_COMMON));
+
+    private static final Integer[] SPACES_ARRAY_AMEX = {4, 11};
+    private static final Set<Integer> SPACE_SET_AMEX =
+            new HashSet<>(Arrays.asList(SPACES_ARRAY_AMEX));
+
+    @VisibleForTesting @Card.CardBrand String mCardBrand = Card.UNKNOWN;
+    private int mLengthMax = 19;
+
+    public CardNumberEditText(Context context) {
+        super(context);
+        listenForTextChanges();
+    }
+
+    public CardNumberEditText(Context context, AttributeSet attrs) {
+        super(context, attrs);
+        listenForTextChanges();
+    }
+
+    public CardNumberEditText(Context context, AttributeSet attrs, int defStyleAttr) {
+        super(context, attrs, defStyleAttr);
+        listenForTextChanges();
+    }
+
+    /**
+     * Updates the selection index based on the current (pre-edit) index, and
+     * the size change of the number being input.
+     *
+     * @param oldPosition the original position of the cursor
+     * @param oldLength the pre-edit length of the string
+     * @param newLength the post-edit length of the string
+     * @return an index within the string at which to put the cursor
+     */
+    @VisibleForTesting
+    int updateSelectionIndex(int oldPosition, int oldLength, int newLength) {
+        int increment, newPosition;
+        Set<Integer> spaceSet = Card.AMERICAN_EXPRESS.equals(mCardBrand)
+                ? SPACE_SET_AMEX
+                : SPACE_SET_COMMON;
+
+        boolean growing = newLength > oldLength;
```

Comment:
Yes -- if you paste over an equal amount of numbers. I'm intentionally not hitting that here, since I'm not yet handling the pasting situation. My evil plan is to just put the cursor at the end of any pasting operation.